File: ex_stay/api/booking.py
import frappe
import json
from frappe import _

@frappe.whitelist(allow_guest=True)
def book(**kwargs):
    try:
        data = frappe._dict(kwargs)
        
        # Validate required fields
        required_fields = [
            "email", "first_name", "last_name", "guest_country",
            "phone", "check_in", "check_out", "no_of_nights",
            "guests", "code", "rate", "price_per_night", "total_price", "nights_x_price_per_night",
            "levies", "vat_frontend"
        ]
        
        missing_fields = [field for field in required_fields if not data.get(field)]
        if missing_fields:
            return {"status": "error", "message": f"Missing required fields: {', '.join(missing_fields)}"}
        
        # Create new booking document
        booking = frappe.get_doc({
            "doctype": "Ex Bookings",
            "email": data.email,
            "first_name": data.first_name,
            "last_name": data.last_name,
            "guest_country": data.guest_country,
            "phone": data.phone,
            "check_in": data.check_in,
            "check_out": data.check_out,
            "no_of_nights": data.no_of_nights,
            "guests": data.guests,
            "code": data.code,
            "rate": data.rate,
            "price_per_night": data.price_per_night,
            "total_price": data.total_price,
            "nights_x_price_per_night": data.nights_x_price_per_night,
            "levies": data.levies,
            "vat_frontend": data.vat_frontend
        })
        
        booking.insert(ignore_permissions=True)
        frappe.db.commit()
        
        return {"status": "success", "message": "Booking created successfully", "booking_id": booking.name}
    
    except Exception as e:
        frappe.log_error(frappe.get_traceback(), "Create Booking API Error")
        return {"status": "error", "message": str(e)}



# http://127.0.0.1:8000/api/method/ex_stay.api.booking.book

File: frontend/src/components/utility/booking_page/booking.js
// src/utils/booking.js
import { ref, computed } from "vue";
import { createResource } from "frappe-ui";
import { session } from '../../../data/session';

export function useBooking(property) { // Accept property as a parameter
  const showReservationForm = ref(false);
  const checkIn = ref("");
  const checkOut = ref("");
  const firstName = ref("");
  const lastName = ref("");
  const country = ref("");
  const telephoneNumber = ref("");
  const guestCount = ref("");
  const selectedCountry = ref("");
  const passportNumber = ref("");
  const selectedPackage = ref("");

  const alertMessage = ref(null);
  const alertType = ref(null);

  const userEmail = computed(() => session.user || "");

  

  const numberOfNights = computed(() => {
    if (checkIn.value && checkOut.value) {
      const start = new Date(checkIn.value);
      const end = new Date(checkOut.value);

      if (isNaN(start) || isNaN(end)) {
        return 0;
      }

      const nights = (end - start) / (1000 * 60 * 60 * 24);
      return nights > 0 ? Math.floor(nights) : 0;
    }
    return 0;
  });

  const formatCurrency = (value) => {
    return new Intl.NumberFormat("en-US", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(value);
  };

  const plusNights = computed(() => {
    if (property.value && property.value.price_list_rate && numberOfNights.value > 0) {
      return formatCurrency(property.value.price_list_rate * numberOfNights.value);
    }
    return formatCurrency(0);
  });

  const isBooking = ref(false);

  const book = async () => {

    if (isBooking.value) return; // Prevent duplicate requests
    isBooking.value = true; // Lock function while request is running


    try {
      console.log("Preparing booking data...");

      // Validation
      if (!firstName.value) {
        alertMessage.value = "⚠️ First Name is required.";
        alertType.value = "warning";
        return;
      }

      if (!lastName.value) {
          alertMessage.value = "⚠️ Last Name is required.";
          alertType.value = "warning";
          return;
      }

      if (!selectedCountry.value) {
          alertMessage.value = "⚠️ Please select a Country.";
          alertType.value = "warning";
          return;
      }

      if (!telephoneNumber.value) {
          alertMessage.value = "⚠️ Phone Number is required.";
          alertType.value = "warning";
          return;
      }

      if (!passportNumber.value) {
          alertMessage.value = "⚠️ Passport Number is required.";
          alertType.value = "warning";
          return;
      }

      if (!checkIn.value) {
          alertMessage.value = "⚠️ Check-in date is required.";
          alertType.value = "warning";
          return;
      }

      if (!checkOut.value) {
          alertMessage.value = "⚠️ Check-out date is required.";
          alertType.value = "warning";
          return;
      }

      if (!guestCount.value || parseInt(guestCount.value) <= 0) {
          alertMessage.value = "⚠️ Please enter a valid number of guests (greater than 0).";
          alertType.value = "warning";
          return;
      }

      if (!selectedPackage.value && property.value?.custom_property_category === 'Hotel') {
        alertMessage.value = "⚠️ Please select a package.";
        alertType.value = "warning";
        return;
    }

      const bookingData = {
        doctype: "Ex Bookings",
        email: userEmail.value || "",
        first_name: firstName.value || "",
        last_name: lastName.value || "",
        guest_country: selectedCountry.value || "",
        phone: telephoneNumber.value || "",
        passport_no: passportNumber.value || "",
        guests: guestCount.value || "",
        check_in: checkIn.value || "",
        check_out: checkOut.value || "",
        no_of_nights: numberOfNights.value,
        code: property.value?.item_code,
        item_price_name:property.value?.name,
        tax_category: property.value?.tax,
        price_per_night: property.value?.price_list_rate,
        property_category: property.value?.custom_property_category,
        hotel_package_selected: selectedPackage.value || "",
      };

      console.log("📩 Booking Payload:", bookingData);

      const bookResponse = await createResource({
        url: 'frappe.client.insert',
        params: {
          doc: bookingData,
        },
      }).fetch();

      console.log("📩 API Response:", bookResponse);

      if (bookResponse && bookResponse.name) {
        console.log("✅ Booking successful!", bookResponse);
        alertMessage.value = `🎉 Booking successful! Booking ID: ${bookResponse.name}`;
        alertType.value = "success";
        showReservationForm.value = false;


        // ✅ Reset fields after successful booking
        checkIn.value = "";
        checkOut.value = "";
        firstName.value = "";
        lastName.value = "";
        selectedCountry.value = "";
        telephoneNumber.value = "";
        passportNumber.value = "";
        guestCount.value = "";
        selectedPackage.value = "";
        
      } else {
        console.error("❌ Booking failed:", bookResponse);
        alertMessage.value = "❌ Booking failed. Please try again.";
        alertType.value = "error";
      }
    } catch (err) {
      console.error("🚨 Unable to book reservation", err);
    }
  };

  return {
    showReservationForm,
    checkIn,
    checkOut,
    passportNumber,
    firstName,
    lastName,
    country,
    telephoneNumber,
    guestCount,
    selectedCountry,
    userEmail,
    numberOfNights,
    formatCurrency,
    plusNights,
    book,
    alertMessage,
    alertType, 
    selectedPackage
  };
}




